lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -0,0 +1,138 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+"use strict";
+
+(function (intrinsic) {
+    var platform = intrinsic.JsBuiltIn;
+
+    let FunctionsEnum = {
+        ArrayValues: { className: "Array", methodName: "values", argumentsCount: 0, forceInline: true /*optional*/, alias: "Symbol.iterator"},
+        ArrayKeys: { className: "Array", methodName: "keys", argumentsCount: 0, forceInline: true /*optional*/ },
+        ArrayEntries: { className: "Array", methodName: "entries", argumentsCount: 0, forceInline: true /*optional*/ }
+    };
+ 
+    var setPrototype = platform.builtInSetPrototype;
+    var _objectDefineProperty = platform.builtInJavascriptObjectEntryDefineProperty;
+
+    // Object's getter and setter can get overriden on the prototype, in that case while setting the value attributes, we will end up with TypeError
+    // So, we need to set the prototype of attributes to null
+    var ObjectDefineProperty = function (obj, prop, attributes) {
+        _objectDefineProperty(obj, prop, setPrototype(attributes, null));
+    };
+    var CreateObject = platform.builtInJavascriptObjectCreate;
+
+    platform.registerChakraLibraryFunction("ArrayIterator", function (arrayObj, iterationKind) {
+        "use strict";  
+        this.arrayObj = arrayObj;
+        this.nextIndex = 0;
+        this.kind = iterationKind;
+    });
+
+    // ArrayIterator's prototype is the C++ Iterator, which is also the prototype for StringIterator, MapIterator etc
+    var iteratorPrototype = platform.GetIteratorPrototype();
+    // Establish prototype chain here
+    __chakraLibrary.ArrayIterator.prototype = CreateObject(iteratorPrototype);
+
+    ObjectDefineProperty(__chakraLibrary.ArrayIterator.prototype, 'next',
+        {
+            value: function () {
+                "use strict";  
+                let o = this;
+
+                if (this === null || this === undefined) {
+                    throw new TypeError("Array.prototype.values: 'this' is null or undefined");
+                }
+
+                if (typeof (o) !== 'object') {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                if (!(o instanceof __chakraLibrary.ArrayIterator)) {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                let a = o.arrayObj;
+                var value, done;
```

Comment:
>var value, done; [](start = 16, length = 16)

initialize this as undefined and true, then you don't need to have it set two places below?

Reply:
Nit: use let instead of var

Reply:
@akroshg The jit will copy prop the values in either cases. I don't want to add an unconditional write which need not happen.